File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int calc(long credit);
void id_credit(long credit);
int valid(int validation_num);

int main(void)
{
    long credit = get_long("Number: "); //prompt user for credit card number

    int validation_num = calc(credit);
    int validation = valid(validation_num);
    if (validation == 0)
    {
        printf("INVALID\n");
    }
    else
    {
        //void id_credit(long credit);
        int count = 0;
        long count_credit = credit;
        int visa_13 = credit/(1e12);
        int amex = credit/(1e13);
        int visa_16 = credit/(1e15);
        int mastercard = credit/(1e14);

        // Get digit count
        while (count_credit > 0)
        {
            count_credit /= 10;
            count++;
        }

        // Check company
        if (count == 13 && visa_13 == 4)
            printf("VISA\n");
        else if (count == 15 && (amex == 34 || amex == 37))
            printf("AMEX\n");
        else if (count == 16 && visa_16 == 4)
            printf("VISA\n");
        else if (count == 16 && (mastercard >= 51 && mastercard <= 55))
            printf("MASTERCARD\n");
        else
            printf("INVALID\n");
    }
}

// Calculate Luhn's Algorithm
int calc(long credit)      // 4003600000000014
{
    int prod_sum = 0;
    long new_credit = credit;
    while (new_credit > 0)
    {
        int digit = new_credit % 100 / 10; //get second to last digit
        int prod_digit = digit * 2;
        if (prod_digit > 9)
        {
            prod_digit = prod_digit % 100 / 10 + prod_digit % 10; //sum double digits
        }
        prod_sum += prod_digit;
        new_credit /= 100; //remove last two digits
    }
    
    int left_sum = 0;
    long new_credit2 = credit;
    while (new_credit2 > 0)
    {
        int left_digit = new_credit2 % 10; //get last number
        left_sum += left_digit;
        new_credit2 /= 100; //remove last two digits
    }

    int validation_num = prod_sum + left_sum;
    //printf("%i\n", validation_num);

    return validation_num;
}

// Determine if card number satisfies Luhn's Algorithm
int valid(int validation_num)
{
    int validation;
    if (validation_num % 10 == 0)
    {
        validation = 1; //VALID
    }
    else
    {
        validation = 0; //INVALID
    }
    return validation;
}

// If card number valid, identify company
void id_credit(long credit)
{
    int count = 0;
    long count_credit = credit;
    int visa_13 = credit/(1e12);
    int amex = credit/(1e13);
    int visa_16 = credit/(1e15);
    int mastercard = credit/(1e14);

    // Get digit count
    while (count_credit > 0)
    {
        count_credit /= 10;
        count++;
    }

    // Check company
    if (count == 13 && visa_13 == 4)
        printf("VISA\n");
    else if (count == 15 && (amex == 34 || amex == 37))
        printf("AMEX\n");
    else if (count == 16 && visa_16 == 4)
        printf("VISA\n");
    else if (count == 16 && (mastercard >= 51 && mastercard <= 55))
        printf("MASTERCARD\n");
    else
        printf("INVALID\n");
}

File: pset6/sentimental-credit/credit.py
# Report if credit card number is valid, if yes identify company
from cs50 import get_int
from math import trunc
import sys


def main():

    # Prompt user for credit card number
    credit = get_int("Number: ")
    credit_list = [int(i) for a, i in enumerate(str(credit))]  # convert number to iterable list

    # Validate card number
    validation(credit_list)

    # Identify company
    id_credit(credit_list)


# Calculate Luhn's Algorithm to validate card number  # test: 4003600000000014
def validation(credit):

    checksum = 0

    # Multiply every other digit by 2 starting from the second to last, add those digits together
    for i in credit[-2::-2]:
        prod_digit = i * 2
        if prod_digit > 9:
            checksum += trunc(prod_digit % 100 / 10) + prod_digit % 10  # sum double digits
        else:
            checksum += prod_digit

    # Add checksum to sum of remaining digits in credit
    for i in credit[-1::-2]:
        checksum += i

    # Card is valid if the checksum is divisible by 10
    if checksum % 10 != 0:
        print("INVALID")
        sys.exit(1)


# Identify company
def id_credit(credit):

    length = len(credit)
    first_num = credit[0]
    second_num = credit[1]

    if length == 13 and first_num == 4:
        print("VISA")
    elif length == 15 and first_num == 3 and (second_num == 4 or second_num == 7):
        print("AMEX")
    elif length == 16 and first_num == 4:
        print("VISA")
    elif length == 16 and first_num == 5 and (second_num >= 1 and second_num <= 5):
        print("MASTERCARD")
    else:
        print("INVALID")
        sys.exit(2)


main()
